app.py

```diff
@@ -12,17 +12,17 @@ def validate(keyboard, pin):
     :return: nothing or raises ValueError if validation fails
     """
     if not all([isinstance(keyboard, list), isinstance(pin, str)]):
-        raise ValueError(
+        raise TypeError(
             f"Input is not valid. Keyboard must be a list. Pin must be a string."
         )
-    if not pin.isdigit():
-        raise ValueError(f"Input is not valid. Pin must be a string of digits.")
-    if not all([isinstance(el, list) for el in keyboard]) or not all(
-        [isinstance(el[i], str) for el in keyboard for i in range(len(el))]
-    ):
-        raise ValueError(
-            f"Input is not valid. Keyboard must be a list of lists of strings."
-        )
+    if keyboard == [] or pin == "":
+        raise ValueError("Keyboard or PIN cannot be empty.")
```

Comment:
There is no added value to merge two errors in a single exception. At worst, when you will get this error (unexpectedly) you will wonder which is it the pin or the keyboard.

Reply:
Roger! Explicit is good :)